### PR TITLE
Added a dslx implementation of a fast fourier transform (fft) based o…

### PR DESCRIPTION
**Pull request by jbaileyhandle**
…n the one from MachSuite (https://breagen.github.io/MachSuite/). Also added a 32-bit floating point subtr act module (trivial modification on top of adder call) and made fp32 add and mul 'pub' functions so they can be imported and used in other files.  Note that the machsuite BSD-3 license is included in the fft directory.

### CODE REVIEW

**Comment by jbaileyhandle** [comment]
User experience note for floating-point modules: floating point is very verbose (eg. fpmul_2x32::fpmul_2x32(x,y)) and it's odd that the floating-point operations are in separate modules. It would be nice if they were all collected and we could do calls like fp:mul_2x32 and fp:add_2x32 etc.  Or better yet if we could overload operators (long term idea - not an immediate priority).

**Comment by jbaileyhandle** [comment]
It seems that fft is too large after unrolling (1000s of nodes at least) for ir optimization to scale well - My desktop (16 GB ram) get bricked when ir optimization runs.  Current 'fix' is just to disable conversion of dslx to ir.

**Comment by meheff** [comment]
I contacted the open source licensing team inside Google and they said the ported code should live in a third_party directory with the original license in the DSLX file. Specifically: google/xls/third_party/machsuite

Inside this directory should be a LICENSE file containing the original MachSuite license. Each file which is ported from machsuite source code should have the *MachSuite* header at the top, not the one we use in XLS. I'd guess the other files in this PR (not the ported ones) can live inside the xls/ directory and be treated just like other code we write. They can just refer to the files in the third_party/machsuite directory for testing purposes etc.

**Comment by jbaileyhandle** [comment]
Thanks for talking to the open source people, Mark.  When we say that the ported files should go under the third_party directory and the non-ported ones are okay where they are, which files would count as ported vs non-ported?  I'd assume fft_reference.cc would count as ported and that the changes to our floating point modules would not (not at all derived from Mach Suite).  I'm fuzzy fft.x and the test data, which follow the MachSuite as a reference but is our own implementation.

**Comment by meheff** [comment]
If in doubt, just put it in third_party (like fft.x). Is the test data a copy of what is in the machsuite? or something you generated on your own, If it's a copy, then put it in third_party.

**Comment by jbaileyhandle** [comment]
Went ahead and moved everything to the third_party directory (I don't think there was anything besides the BUILD file that wasn't at least partly derivative).

File: xls/third_party/machsuite/fft/fft_reference.cc
Comment (meheff) [inline xls/third_party/machsuite/fft/fft_reference.cc:116]: union float_unit_union {
...
};

File: xls/third_party/machsuite/fft/fft_reference.cc
Comment (meheff) [inline xls/third_party/machsuite/fft/fft_reference.cc:131]: std::cout << absl::StreamFormat("0x%x,\n", data_union.u32);

**Review by meheff — CHANGES_REQUESTED** [review]
Thanks for adding this!

File: xls/third_party/machsuite/fft/fft_reference.cc
Comment (jbaileyhandle) [reply xls/third_party/machsuite/fft/fft_reference.cc:116]: uint as in uint32_t.

File: xls/third_party/machsuite/fft/fft_reference.cc
Comment (jbaileyhandle) [reply xls/third_party/machsuite/fft/fft_reference.cc]: Done.

File: xls/third_party/machsuite/fft/fft_reference.cc
Comment (meheff) [reply xls/third_party/machsuite/fft/fft_reference.cc:117]: I meant change the form of the declaration not the name. That is:

union foo {
};

Not:

typedef union foo_t {
} foo;

I'll make the change as it's uploaded internally.

**Comment by jbaileyhandle** [comment]
Is the auto-puller (don't know real name) getting hung up on something?

**Comment by meheff** [comment]
fyi, we're just waiting on one last approval before this can land internally and mirrored in github. It had to be split into several different changes, so not sure how that will be reflected in this PR after it lands.

**Comment by meheff** [comment]
This change has landed but had to be broken into multiple bits so it's not reflected in the PR state. This can be reverted/withdrawn or whatever one does on Gtihub.

**Pull request merged**
